### ISSUE TITLE
Disk I/O Error Intermittently after deleting and recreating the database

### ISSUE THREAD

**Issue by pasin** [couchbase/couchbase-lite-core#539]
This has been found only from running unit tests on Jenkins which is slow. The unit test that can cause this issue is [ReplicatorTest.testShortP2P](https://github.com/couchbase/couchbase-lite-ios/blob/master/Objective-C/Tests/ReplicatorTest.m#L1029-L1089) which run the test twice and delete and reopen the database after each run.

**Log:** 

I/O Error Case:
https://gist.github.com/pasin/54c0fa06a0669560990ec1caade9eb98#file-io_error-txt

Success Case:
https://gist.github.com/pasin/54c0fa06a0669560990ec1caade9eb98#file-no_io_error-txt

**Comment by pasin** [comment]
Captured from the log linked above,

```
17:30:05 2018-07-17 17:29:56.767931-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: Preparing to delete DataFile /private/var/folders/vt/_5s5821s1g7fdb0z04jyjgm40000gp/T/CouchbaseLite_EE/otherdb.cblite2/db.sqlite3
17:30:05 2018-07-17 17:29:56.768153-0700 xctest[36327:1616049] CouchbaseLite DB Debug: File 0x612000ab73c0: Remove DataFile 0x611000aa0a40
17:30:05 2018-07-17 17:29:56.768241-0700 xctest[36327:1616049] CouchbaseLite DB Info: Closing database /var/folders/vt/_5s5821s1g7fdb0z04jyjgm40000gp/T/CouchbaseLite_EE/otherdb.cblite2/db.sqlite3
17:30:05 2018-07-17 17:29:56.768626-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: Pre-close housekeeping: 0 of 9 pages free (0%)
17:30:05 2018-07-17 17:29:56.770999-0700 xctest[36327:1616049] CouchbaseLite DB Info: Deleting database file /private/var/folders/vt/_5s5821s1g7fdb0z04jyjgm40000gp/T/CouchbaseLite_EE/otherdb.cblite2/db.sqlite3 (with -wal and -shm)
17:30:05 2018-07-17 17:29:56.772225-0700 xctest[36327:1616049] CouchbaseLite DB Info: Opening CBLDatabase[otherdb] at path /var/folders/vt/_5s5821s1g7fdb0z04jyjgm40000gp/T/CouchbaseLite_EE/otherdb.cblite2
17:30:05 2018-07-17 17:29:56.772795-0700 xctest[36327:1616049] CouchbaseLite DB Debug: File 0x612000ab73c0: adding DataFile 0x611000b38ac0
17:30:05 2018-07-17 17:29:56.772892-0700 xctest[36327:1616049] CouchbaseLite DB Info: Opening database /var/folders/vt/_5s5821s1g7fdb0z04jyjgm40000gp/T/CouchbaseLite_EE/otherdb.cblite2/db.sqlite3
17:30:05 2018-07-17 17:29:56.776646-0700 xctest[36327:1616049] CouchbaseLite DB Debug: DataFile: open KVS 'default'
17:30:05 2018-07-17 17:29:56.778272-0700 xctest[36327:1616049] CouchbaseLite DB Debug: DataFile: open KVS 'info'
17:30:05 2018-07-17 17:29:56.779989-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: DataFile: begin transaction
17:30:05 2018-07-17 17:29:56.780183-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(info) set versioning
17:30:05 2018-07-17 17:29:56.780689-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(info) set publicUUID
17:30:05 2018-07-17 17:29:56.780876-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(info) set privateUUID
17:30:05 2018-07-17 17:29:56.781018-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: DataFile: commit transaction
17:30:05 2018-07-17 17:29:56.781572-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: DataFile: begin transaction
17:30:05 2018-07-17 17:29:56.781980-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(default) insert livesindb
17:30:05 2018-07-17 17:29:56.782319-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(info) set SharedKeys
17:30:05 2018-07-17 17:29:56.782469-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: DataFile: commit transaction
17:30:05 2018-07-17 17:29:56.783390-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: DataFile: begin transaction
17:30:05 2018-07-17 17:29:56.783742-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(default) insert livesinotherdb
17:30:05 2018-07-17 17:29:56.784066-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: KeyStore(info) set SharedKeys
17:30:05 2018-07-17 17:29:56.784213-0700 xctest[36327:1616049] CouchbaseLite DB Verbose: DataFile: commit transaction
17:30:05 2018-07-17 17:29:56.785296-0700 xctest[36327:1616049] CouchbaseLite Sync Info: CBLReplicator[-> <CBLMessageEndpoint: 0x604000464150>]: Starting
17:30:05 2018-07-17 17:29:56.785814-0700 xctest[36327:1616049] CouchbaseLite DB Debug: File 0x612000ab2140: adding DataFile 0x611000b04900
17:30:05 2018-07-17 17:29:56.785901-0700 xctest[36327:1616049] CouchbaseLite DB Info: Opening database /var/folders/vt/_5s5821s1g7fdb0z04jyjgm40000gp/T/CouchbaseLite_EE/testdb.cblite2/db.sqlite3
17:30:05 2018-07-17 17:29:56.786607-0700 xctest[36327:1616049] CouchbaseLite  WARNING: SQLite error (code 522): disk I/O error
17:30:05 2018-07-17 17:29:56.787004-0700 xctest[36327:1616049] CouchbaseLite DB Debug: DataFile: destructing (~DataFile) 0x611000b04900
17:30:05 2018-07-17 17:29:56.787090-0700 xctest[36327:1616049] CouchbaseLite DB Debug: File 0x612000ab2140: Remove DataFile 0x611000b04900
```

It seemed like the database file didn't get deleted immediately but was deleted when the replicator was started?

**Comment by borrrden** [comment]
This looks exactly like https://github.com/couchbase/couchbase-lite-core/issues/381 ..... why is this back again!?

**Comment by pasin** [comment]
I just updated the log files for comparing between I/O error case and success case. I think one of the main differences is that for the I/O error case the database file was deleted while the DBWorker, Repl, and Pull were still working on cleaning tasks.

**Comment by borrrden** [comment]
The database file is not allowed to be deleted at that time, and any attempts should be throwing an exception since there is still an open connection to it (assuming it goes through the C4 API).  The issue the last time this happened was that there was a bug in SQLite regarding zombie mode that would delete db files even after they didn't belong to them anymore.  I don't think that can happen anymore but without a reproduction case this is hard to reason about.

**Comment by borrrden** [comment]
Also does this go away with the addition of the delay in the test?

**Comment by pasin** [comment]
Adding delay seems to help getting things to work in order and made all the tests pass on Jenkins.

**Comment by snej** [comment]
I noticed a discrepancy in the paths -- sometimes the path starts with `/private/var/folders/...` and other times `/var/folders/vt`. Those are equivalent because `/var` is a symlink, but the code that tracks multiple DB instances on the same file relies on the pathname matching. So something may be wrong here.

**Comment by snej** [comment]
I've made some improvements to logging that will help distinguish different database instances; in the logs posted above you can't really tell which line is logged by which.
Can we take out the delay workaround, and examine the next error in Jenkins?

**Comment by pasin** [comment]
OK, I will do that.

**Comment by snej** [comment]
There was a bug I fixed last month (in commit c8db29f), that sometimes caused a database to be left open, which could then cause an error when trying to delete the database. That might be what was happening here.

@pasin, have we retried this test without the workaround?

**Comment by pasin** [comment]
I have removed the sleep workaround and fixed the unit test's setup/teardown to make sure to use Database.delete() to delete the database instead of using FileManager to delete. After that I haven't seen the Disk IO error. I think we can close the issue now.
